### PR TITLE
Adding functionality to initialise RunReaders at offset

### PR DESCRIPTION
**Pull request by recursion-ninja**
# Description

First part of Work Product 12, adding the functionality to start `RunReaders` at an offset.

The new `rawPageFindKey` function has the same performance characteristics as `rawPageLookup` but facilitates starting the `RunReaders` at an give page offset from the returned index. `RunReader.new` now takes an optional offset as an argument.

### CODE REVIEW

File: test/Test/Database/LSMTree/Internal/RawPage.hs [hunk 1]
```diff
@@ -294,7 +296,7 @@ prop_big_insert k mblobref =
   where
     v       = Ref.Value (BS.replicate 5000 42)
     op      = Ref.Insert v mblobref
-    kops    = [(k, op)]
+    kops    = [ (k, op)]
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Internal/RawPage.hs:299]:
```suggestion
    kops    = [(k, op)]
```

File: test/Test/Database/LSMTree/Internal/RawPage.hs [hunk 2]
```diff
@@ -187,6 +187,8 @@ tests = testGroup "Database.LSMTree.Internal.RawPage"
     , testProperty "entry" prop_single_entry
     , testProperty "rawPageOverflowPages" prop_rawPageOverflowPages
     , testProperty "from/to reference impl" prop_fromToReferenceImpl
+    , testProperty "∀ key page. maybe True (key <=) (getRawPageIndexKey . rawPageIndex page .  id  =<< rawPageFindKey page key)" prop_findKey_index_law_GT
+    , testProperty "∀ key page. maybe True (key > ) (getRawPageIndexKey . rawPageIndex page . pred =<< rawPageFindKey page key)" prop_findKey_index_law_LT
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Internal/RawPage.hs:191]: One argument in favour of simpler test names, e.g. just `prop_findKey_index_law_LT`, is that it's easier to search the code for these names, and it's easier to run the tests in isolation using `-p`. What are your thoughts?

I often just put the name of the function in the tests, and if I want to describe what a test does more succinctly then I'll put it as a haddock comment on the function

File: src/Database/LSMTree/Internal/RunReader.hs [hunk 3]
```diff
@@ -187,10 +209,28 @@ next fs hbio reader@RunReader {..} = do
...
+pageNoToAbsOffset :: Index.PageNo -> FS.AbsOffset
+pageNoToAbsOffset (Index.PageNo n) =
+    assert (n >= 0) $
+      FS.AbsOffset (mulPageSize (fromIntegral n))
+
 -- | Returns 'Nothing' on EOF.
 readDiskPage :: HasFS IO h -> FS.Handle h -> IO (Maybe RawPage)
 readDiskPage fs h = do
     mba <- newPinnedByteArray pageSize
+    -- TODO: make sure no other exception type can be thrown
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal/RunReader.hs:233]: I don't think you can prevent this

File: src/Database/LSMTree/Internal/RunReader.hs [hunk 4]
```diff
@@ -64,18 +65,39 @@ data RunReader s fhandle = RunReader {
...
+        -- load first page from disk, if it exists.
+        firstPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs pageNo readerKOpsHandle
+
+        case rawPageFindKey firstPage offset of
+          Just n ->
+            return (firstPage, n)
+          Nothing -> do
+            -- actually, it's the next page
+            nextPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs pageNo readerKOpsHandle
+            return (nextPage, 0)
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal/RunReader.hs:94]: We use `readDiskpage` in `next`, which means that we'll read the first page at an offset, but then start reading pages from the start. This is probably not caught by tests because we never start reading at an offset. We should extend the tests we already have to ensure we're testing this case properly

File: src/Database/LSMTree/Internal/RawPage.hs [hunk 5]
```diff
@@ -186,6 +246,18 @@ data RawPageIndex entry =
      | IndexEntryOverflow !SerialisedKey !entry !Word32
   deriving stock (Eq, Functor, Show)
 
+-- |
+-- __Time:__ \( \mathcal{O}\left( 1 \right) \)
+--
+-- Conveniently access the the 'SerialisedKey' of a 'RawPageIndex'.
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal/RawPage.hs:252]:
```suggestion
-- Conveniently access the 'SerialisedKey' of a 'RawPageIndex'.
```

**Review by jorisdral — COMMENTED** [review]
This is not a full review yet -- here are some initial comments

**Review by jorisdral — CHANGES_REQUESTED** [review]
Whoops, forgot to request changes

File: src/Database/LSMTree/Internal/RunReader.hs (hunk 4 above)
Comment (mheinzel) [reply src/Database/LSMTree/Internal/RunReader.hs:94]: Ah, yes, the handle still points at the beginning, because `hGetBufExactlyAt` does not change the handle state. Maybe we should use `hSeek` instead?

Alternatively, we could keep track of the page number ourselves and always use `hGetBufExactlyAt`.

**Comment by mheinzel** [comment]
There is a [build failure for GHC versions before 9.6](https://github.com/IntersectMBO/lsm-tree/actions/runs/10493900676/job/29068935700?pr=346#step:13:151) which seems related to `UnboxedSums`.

Also, there is a [docspec-related error](https://github.com/IntersectMBO/lsm-tree/actions/runs/10493900676/job/29068937398?pr=346#step:17:103), seems to happen on all Ubuntu builds.

File: src/Database/LSMTree/Internal/RawPage.hs [hunk 6]
```diff
@@ -153,26 +165,74 @@ rawPageLookup !page !key
       | otherwise
       = LookupEntryNotPresent
 
+-- |
+-- __Time:__ \( \mathcal{O}\left( \log_2 n \right) \)
+-- where \( n \) is the number of entries in the 'RawPage'.
+--
+-- Return the least entry number in the 'RawPage' (if it exists)
+-- which is greater than or equal to the suppled 'SerialisedKey'.
```
Comment (mheinzel) [inline src/Database/LSMTree/Internal/RawPage.hs:173]:
```suggestion
-- which is greater than or equal to the supplied 'SerialisedKey'.
```

File: test/Test/Database/LSMTree/Internal/RawPage.hs (hunk 2 above)
Comment (mheinzel) [inline test/Test/Database/LSMTree/Internal/RawPage.hs:191]: These tests only assert anything when `rawPageFindKey` returns `Just`. Can you also add a property that whenever the last key of the page is `(>= key)`, then we successfully find an index number?

Looking at the code I have a hunch that it might sometimes incorrectly return `Nothing`, but I could be wrong.

**Review by mheinzel — CHANGES_REQUESTED** [review]

File: test/Test/Database/LSMTree/Internal/RawPage.hs (hunk 2 above)
Comment (recursion-ninja) [reply test/Test/Database/LSMTree/Internal/RawPage.hs:191]: In my own projects, for type-class laws, I have included the very verbose name of the law. However, I think that since this project has consistently used shorter test names I should modify these test names to follow suit. 

I think I'd like to, if you don't object, move the verbose "relational law" into the counter example message which appears only if the property fails to hold.

File: src/Database/LSMTree/Internal/RunReader.hs [hunk 7]
```diff
@@ -187,10 +209,28 @@ next fs hbio reader@RunReader {..} = do
   Utilities
 -------------------------------------------------------------------------------}
 
+-- | Returns 'Nothing' on EOF.
+readDiskPageAt :: HasFS IO h -> Index.PageNo -> FS.Handle h -> IO (Maybe RawPage)
+readDiskPageAt fs pageNo h = do
+    mba <- newPinnedByteArray pageSize
+    -- TODO: make sure no other exception type can be thrown
```
Comment (dcoutts) [inline src/Database/LSMTree/Internal/RunReader.hs:216]: Others will just propagate. Is that ok?

File: src/Database/LSMTree/Internal/RawPage.hs [hunk 8]
```diff
@@ -153,26 +165,74 @@ rawPageLookup !page !key
...
+      | key <= rawPageKeyAt page 0 = Just 0
+      | otherwise                  = Nothing
+
+-- | This data-type facilitates the code reuse of 'bisectPageToKey' between the
+-- and 'rawPageLookup' and 'rawPageFindKey'.
+--
+-- NOTE: Be sure to enable UnboxedSums to improve efficiency of this data-type.
+data BinarySearchResult
+  = KeyNotInPage
+  | KeyFoundAt {-# UNPACK #-} !Ordering  {-# UNPACK #-} !Int
```
Comment (dcoutts) [inline src/Database/LSMTree/Internal/RawPage.hs:201]: The warnings here are correct, you cannot unpack the `Ordering`, but on the other hand given that it's an enum, there's little advantage either.

Note also that we only pattern match on EQ or not EQ, so we don't need the full `Ordering`. A bool would be enough.

File: src/Database/LSMTree/Internal/RawPage.hs [hunk 9]
```diff
@@ -153,26 +165,74 @@ rawPageLookup !page !key
...
+    !dirNumKeys = rawPageNumKeys page
+
+    lookup1
+      | key <= rawPageKeyAt page 0 = Just 0
+      | otherwise                  = Nothing
+
+-- | This data-type facilitates the code reuse of 'bisectPageToKey' between the
+-- and 'rawPageLookup' and 'rawPageFindKey'.
+--
+-- NOTE: Be sure to enable UnboxedSums to improve efficiency of this data-type.
```
Comment (dcoutts) [inline src/Database/LSMTree/Internal/RawPage.hs:198]: Is this a TODO? Enabling `UnboxedSums` will not do anything automatically, you'd have to explicitly make `bisectPageToKey` return an unboxed sum. There's no worker/wrapper equivalent to do this automagically, AFAIK.

File: src/Database/LSMTree/Internal/RawPage.hs [hunk 10]
```diff
@@ -153,26 +165,74 @@ rawPageLookup !page !key
...
-            EQ -> LookupEntry (rawPageEntryAt page k)
-            GT -> bisect (k + 1) j
-            LT -> bisect i k
-      where
-        k = i + div2 (j - i)
-
-    linear :: Int -> Int -> RawPageLookup (Entry SerialisedValue BlobSpan)
+    -- k in [i, j)
+    -- k >= key
+    -- k-1 < key
```
Comment (dcoutts) [inline src/Database/LSMTree/Internal/RawPage.hs:228]: Presumably these are preconditions? Say so explicitly.

File: test/Test/Database/LSMTree/Internal/Run.hs [hunk 11]
```diff
@@ -262,7 +262,7 @@ isLargeKOp (key, entry) = size > pageSize
 
 readKOps :: FS.HasFS IO h -> FS.HasBlockIO IO h -> Run RealWorld (FS.Handle h) -> IO [SerialisedKOp]
 readKOps fs hbio run = do
-    reader <- Reader.new fs hbio run
+    reader <- Reader.new fs hbio Nothing run
```
Comment (dcoutts) [inline test/Test/Database/LSMTree/Internal/Run.hs:265]: Will we have tests for the new functionality later? :-) In this PR or another PR?

**Review by dcoutts — CHANGES_REQUESTED** [review]
This is looking good. No major grumbles.

File: src/Database/LSMTree/Internal/RunReader.hs (hunk 3 above)
Comment (recursion-ninja) [reply src/Database/LSMTree/Internal/RunReader.hs:233]: Yeah, I put that note in more as a reminder to review this area later.

File: test/Test/Database/LSMTree/Internal/RawPage.hs (hunk 2 above)
Comment (recursion-ninja) [reply test/Test/Database/LSMTree/Internal/RawPage.hs:191]: @mheinzel Your hunch seems to have been accurate. I added a property test for the `Nothing` case and it does fail! Any thoughts on when it incorrectly returns `Nothing`?

File: src/Database/LSMTree/Internal/RunReader.hs (hunk 7 above)
Comment (recursion-ninja) [reply src/Database/LSMTree/Internal/RunReader.hs:216]: After having some weeks to think about it, I don't think we can properly handle other exceptions at this call site.

Comment (recursion-ninja) [reply src/Database/LSMTree/Internal/RunReader.hs]: @dcoutts I have added a paraphrased version of your commentary as a comment to `RunReader.new`.

File: test/Test/Database/LSMTree/Internal/RawPage.hs (hunk 2 above)
Comment (recursion-ninja) [reply test/Test/Database/LSMTree/Internal/RawPage.hs]: You can replicate a failure case on this branch by invoking:

```
cabal test lsm-tree-test --test-options='--quickcheck-replay="(SMGen 3937336630516587970 3943216271980923571,0)" -p "/rawPageFindKey missing/"'
```

File: test/Test/Database/LSMTree/Internal/Run.hs (hunk 11 above)
Comment (recursion-ninja) [reply test/Test/Database/LSMTree/Internal/Run.hs:265]: Another PR I think. I'll need some guidance on the semantics of how to set up an "imperative" test correctly with our helper functions.

File: src/Database/LSMTree/Internal/RunReaders.hs [hunk 12]
```diff
@@ -91,7 +92,26 @@ new fs hbio runs = do
       return Readers {..}
   where
     fromRun :: ReaderNumber -> Run RealWorld (FS.Handle h) -> IO (Maybe (ReadCtx (FS.Handle h)))
-    fromRun n run = nextReadCtx fs hbio n =<< Reader.new fs hbio run
+    fromRun n run = nextReadCtx fs hbio n =<< Reader.new fs hbio Nothing run
+
+-- | On equal keys, elements from runs earlier in the list are yielded first.
+-- This means that the list of runs should be sorted from new to old.
+-- TODO: merge with 'new'?
+newAtOffset :: forall h .
```
[the remark this answers is not in the document]
Comment (recursion-ninja) [reply src/Database/LSMTree/Internal/RunReaders.hs:102]: I transferred the functionality to a generalized `newAtOffsetMaybe` function. Both `new` and `newAtOffset` call `newAtOffsetMaybe` f while being able to retain their original type signatures (no `Maybe` in `newAtOffset`).

File: test/Test/Database/LSMTree/Internal/RawPage.hs (hunk 2 above)
Comment (recursion-ninja) [reply test/Test/Database/LSMTree/Internal/RawPage.hs]: @mheinzel I have figured out the corner case in the binary search!

File: src/Database/LSMTree/Internal/RunReader.hs [hunk 13]
```diff
@@ -71,26 +71,41 @@ new ::
...
+          -- it would be between the last key in this page and the first key in the next page.
+          -- Thus the reader should be initialised to return keys starting from the next page.
+          Nothing | pageNo == pageEnd ->
+            -- If already at the last page, there is no page of content to return.
+            -- Hence, no need to call readDiskPageAt.
+            return (emptyRawPage, 0)
+
+          _ -> do
+            -- Otherwise, we return the beginning of the next page.
+            nextPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs (Index.nextPageNo pageNo) readerKOpsHandle
```
Comment (dcoutts) [inline src/Database/LSMTree/Internal/RunReader.hs:108]: So this tries to handle the case where there is no next page, because we were already at the last page (corresponding to asking for a greater than all other keys).

I see that `readDiskPageAt` looks for the `FsReachedEOF` exception. Do we know it definitely works that way when using the "at" functions, like `hGetBufExactlyAt` to request something past the end of the file?

I think it's worth adding a unit test (or a label in a property test) to make sure we cover the case of asking for a key that's past the end of all the keys in the db, so we cover this case here.

File: src/Database/LSMTree/Internal/RunReaders.hs [hunk 14]
```diff
@@ -86,15 +86,7 @@ new :: forall h .
...
-new fs hbio runs = do
-    readers <- zipWithM (fromRun . ReaderNumber) [1..] runs
-    for (nonEmpty (catMaybes readers)) $ \xs -> do
-      (readersHeap, readCtx) <- Heap.newMutableHeap xs
-      readersNext <- newMutVar readCtx
-      return Readers {..}
-  where
-    fromRun :: ReaderNumber -> Run IO (FS.Handle h) -> IO (Maybe (ReadCtx (FS.Handle h)))
-    fromRun n run = nextReadCtx fs hbio n =<< Reader.new fs hbio Nothing run
+new = newAtOffsetMaybe Nothing
```
Comment (dcoutts) [inline src/Database/LSMTree/Internal/RunReaders.hs:89]: :+1:

**Review by dcoutts — APPROVED** [review]
LGTM.

Once you're happy enough reviews are in, squash these commits together.

File: test/Test/Database/LSMTree/Internal/RawPage.hs (hunk 2 above)
Comment (mheinzel) [reply test/Test/Database/LSMTree/Internal/RawPage.hs]: Nice!

The properties we seem a little complex by now. On the other hand, it is quite straightforward to to create a reference implementation working on the list of k/ops, similar to the test for `rawPageLookup`. Maybe that would be a nicer approach?

File: src/Database/LSMTree/Internal/RawPage.hs [hunk 15]
```diff
@@ -153,26 +164,90 @@ rawPageLookup !page !key
...
+
+    lookup1
+      | key <= rawPageKeyAt page 0 = Just 0
+      | otherwise                  = Nothing
+
+-- | This data-type facilitates the code reuse of 'bisectPageToKey' between the
+-- and 'rawPageLookup' and 'rawPageFindKey'.
+data BinarySearchResult
+  = KeyNotInPage
+  | KeyFoundAt !Bool {-# UNPACK #-} !Int
```
Comment (mheinzel) [inline src/Database/LSMTree/Internal/RawPage.hs:202]: The Bool here is not explained and generally I think the constructor names are a bit misleading. `KeyNotInPage` means specifically that the key is larger than the largest key in the page. `KeyFoundAt False i` means that the key was *not* found, but the next-largest key can be found at index `i`.

Maybe three constructors would be clearer? (I think the allocation should still be optimised away, but might be worth checking)

```haskell
  = KeyIsTooLarge
  | KeyNotFoundNextIs {-# UNPACK #-} !Int
  | KeyFoundAt {-# UNPACK #-} !Int
```

We can also leave it at two constructors, but then tweaking the name and commenting generously would be nice.

File: src/Database/LSMTree/Internal/RawPage.hs [hunk 16]
```diff
@@ -153,26 +164,90 @@ rawPageLookup !page !key
...
+    -- If there is another key after the linear scan area,
+    -- then we must perform a final comparison to ensure that the key of boundary index is not
+    -- the smallest key which is greater than the search key.
+    {-# INLINE linear #-}
+    linear :: Int -> Int -> BinarySearchResult
     linear !i !j
-        | i >= j                     = LookupEntryNotPresent
-        | key == rawPageKeyAt page i = LookupEntry (rawPageEntryAt page i)
-        | otherwise                  = linear (i + 1) j
+      | i > j || (i == j && i == quantity) = KeyNotInPage
```
Comment (mheinzel) [inline src/Database/LSMTree/Internal/RawPage.hs:246]: `i > j` should never happen, so no need to check it. We could add `i <= j` as an invariant in the comment above and also an assertion.

Also, if `i == j` but `i /= quantity`, we know immediately that `KeyFoundAt False j` should be returned (the range `[i, j)` is empty), no need to call `rawPageKeyAt` again.

So overall this might look like `| i == j = if i == quantity then KeyNotInPage else KeyFoundAt False j`.

File: src/Database/LSMTree/Internal/RawPage.hs [hunk 17]
```diff
@@ -153,26 +164,90 @@ rawPageLookup !page !key
...
+
+-- | Binary search procedure shared between 'rawPageLookup' and 'rawPageFindKey'.
+{-# INLINE bisectPageToKey #-}
+bisectPageToKey
+  :: RawPage
+  -> SerialisedKey
+  -> BinarySearchResult
+bisectPageToKey !page !key = go 0 quantity
+  where
+    quantity = fromIntegral $ rawPageNumKeys page
```
Comment (mheinzel) [inline src/Database/LSMTree/Internal/RawPage.hs:212]: Something like `numKeys` would make it clearer which quantity is being referred to.

File: src/Database/LSMTree/Internal/RunReader.hs (hunk 13 above)
Comment (mheinzel) [reply src/Database/LSMTree/Internal/RunReader.hs:108]: I would also like to see some tests to be confident about the correctness. It's a bit tricky.

For example, I think `nextPageNo pageNo` here doesn't work if there is a multi-page value. We'd end up loading an overflow page. It should be `nextPageNo pageEnd` instead.

File: src/Database/LSMTree/Internal/RunReader.hs [hunk 18]
```diff
@@ -64,18 +65,54 @@ data RunReader m fhandle = RunReader {
...
+          -- but then rawPageFindKey tells us that in fact
+          -- there is no key greater than or equal to the given offset on this page.
+          -- So what can we conclude from the given information?
+          -- The information tells us that the key does not exist, but that if it were to exist,
+          -- it would be between the last key in this page and the first key in the next page.
+          -- Thus the reader should be initialised to return keys starting from the next page.
+          Nothing | pageNo == pageEnd ->
+            -- If already at the last page, there is no page of content to return.
+            -- Hence, no need to call readDiskPageAt.
+            return (emptyRawPage, 0)
```
Comment (mheinzel) [inline src/Database/LSMTree/Internal/RunReader.hs:104]: I don't understand this case. Do you mean something like `pageNo >= sizeInPages index`? 

`pageEnd` is not the last page of the entire run, just the last page of the (potential) entry we are looking for. So `pageNo == pageEnd` means that we don't have to deal with a larger-than-page entry here.

File: src/Database/LSMTree/Internal/RunReader.hs (hunk 4 above)
Comment (mheinzel) [reply src/Database/LSMTree/Internal/RunReader.hs]: > you're going to have to either systematically use `readDiskPageAt` or use seeking

I don't think this has been resolved yet @recursion-ninja.

> Using `readDiskPageAt` and keeping track ourselves would get us closer to being able to use the same `O_DIRECT` file handle as the lookups use (we'd have to add our own manual read-ahead)

Do we want this? We save a file handle, but creating a new one for sequential reading and simply relying on the OS to do read-ahead seems quite convenient.

**Review by mheinzel — COMMENTED** [review]
Still a few things that should be addressed IMO, mainly the issues around keys that fall within pages (where `rawPageFindKey` returns `Nothing`) and the usage of `readDiskPage` which currently doesn't take the offset into account properly.

File: src/Database/LSMTree/Internal/RunReader.hs (hunk 13 above)
Comment (mheinzel) [reply src/Database/LSMTree/Internal/RunReader.hs:108]: One way to get some coverage would be to add `newAtOffset` to the tests in `Test.Database.LSMTree.Internal.RunReaders`.

Comment (dcoutts) [reply src/Database/LSMTree/Internal/RunReader.hs:108]: > For example, I think nextPageNo pageNo here doesn't work if there is a multi-page value.

Ah yes, you're right.

File: src/Database/LSMTree/Internal/RawPage.hs (hunk 15 above)
Comment (recursion-ninja) [reply src/Database/LSMTree/Internal/RawPage.hs]: I expanded the sum type to be more explicit.

```haskell
data BinarySearchResult                                                                                                         
  = KeyNotFoundExceedsPage                                                                                                      
  | KeyNotFoundNearestIsAt {-# UNPACK #-} !Int                                                                                  
  | KeyWasFoundAt {-# UNPACK #-} !Int 
```

File: src/Database/LSMTree/Internal/RawPage.hs (hunk 16 above)
Comment (recursion-ninja) [reply src/Database/LSMTree/Internal/RawPage.hs]: You are correct, that is equivalent and simpler logic. I will leave the `i >= j` for the sake of ensuring totality of the stopping condition though. You never know when a cosmic ray might flip a bit and we need the loop to stop 😉.

File: src/Database/LSMTree/Internal/RunReader.hs (hunk 13 above)
Comment (recursion-ninja) [reply src/Database/LSMTree/Internal/RunReader.hs:104]: Ah, I understand better now. I changed the parameter to `Index.nextPageNo pageEnd`.

Just to be certain, it is *unconditionally* `Index.nextPageNo pageEnd`, or do we need to check for a multi-page value and branch between `Index.nextPageNo pageEnd` and `Index.nextPageNo pageNo`?

File: src/Database/LSMTree/Internal/RawPage.hs (hunk 17 above)
Comment (recursion-ninja) [reply src/Database/LSMTree/Internal/RawPage.hs:214]: Done!

File: src/Database/LSMTree/Internal/RunReader.hs (hunk 13 above)
Comment (mheinzel) [reply src/Database/LSMTree/Internal/RunReader.hs:104]: Unconditionally, since for single-page values `pageNo == pageEnd`.

File: src/Database/LSMTree/Internal/RawPage.hs (hunk 15 above)
Comment (mheinzel) [reply src/Database/LSMTree/Internal/RawPage.hs]: `Next` would be a bit clearer than `Nearest`. The previous key might be very "near", but we still return the next one.

File: src/Database/LSMTree/Internal/RunReader.hs (hunk 4 above)
Comment (mheinzel) [reply src/Database/LSMTree/Internal/RunReader.hs]: This has been resolved by seeking, so we still rely on the handle's internal file pointer.

File: test/Test/Database/LSMTree/Internal/RawPage.hs [hunk 19]
```diff
@@ -314,3 +318,175 @@ prop_single_entry (Ref.PageContentSingle k op) =
...
+        let keys = fst <$> xs
+            keySmall = Ref.toSerialisedKey $ minimum keys
+            keyLarge = Ref.toSerialisedKey $ maximum keys
+
+            missingWithinRange
+              | keySmall == keyLarge = []
+              | otherwise =
+                let missingKey = do
+                      let n = length keys - 1
+                      afterIndex <- (`mod` n) <$> arbitrary
```
Comment (mheinzel) [inline test/Test/Database/LSMTree/Internal/RawPage.hs:434]: Seems like a job for [chooseInt](https://www.stackage.org/haddock/lts-22.8/QuickCheck-2.14.3/Test-QuickCheck.html#v:chooseInt).

File: test/Test/Database/LSMTree/Internal/RawPage.hs [hunk 20]
```diff
@@ -314,3 +318,175 @@ prop_single_entry (Ref.PageContentSingle k op) =
...
+              , (1, arbitrarySerialisedKeyLessThan keySmall)
+              , (1, arbitrarySerialisedKeyMoreThan keyLarge)
+              ] <> missingWithinRange
+
+-- |
+-- Generate an arbitrary 'SerialisedKey' between the given two keys.
+--
+-- The first key is assumed to be the "smaller" of the pair.
+--
+-- NOTE: Under pathological input conditions, there is an infintesimally small chance
```
Comment (mheinzel) [inline test/Test/Database/LSMTree/Internal/RawPage.hs:451]:
```suggestion
-- NOTE: Under pathological input conditions, there is an infinitesimally small chance
```

File: test/Test/Database/LSMTree/Internal/RawPage.hs [hunk 21]
```diff
@@ -314,3 +318,175 @@ prop_single_entry (Ref.PageContentSingle k op) =
...
+    SerialisedKey . RawBytes <$> VP.zipWithM withinByte loBytes hiBytes
+  where
+    withinByte !lhs !rhs
+      | lhs == rhs = pure rhs
+      | otherwise =
+        let (!lo, !hi)
+              | lhs < rhs = (lhs, rhs)
+              | otherwise = (rhs, lhs)
+            !diff = hi - lo
+        in  (succ lo +) . (`mod` diff) <$> arbitrary
```
Comment (mheinzel) [inline test/Test/Database/LSMTree/Internal/RawPage.hs:465]: This does a slightly weird thing, where *each byte* is between the corresponding bytes of the two keys. This limits the set of keys that get generated quite a bit. If the first byte is already smaller/bigger than the first byte of the lower/upper bound, then the remaining bytes could be completely arbitrary.

Comment (mheinzel) [reply test/Test/Database/LSMTree/Internal/RawPage.hs:465]: We can leave this as is, since the only thing observed about keys is their ordering towards each other. Still feels a bit off.

A potential solution would be to use a simpler generator (maybe only 50:50 existing key or arbitrary key), with labels for the cases specifically targetted here, to make sure the distribution looks good. Recognizing a property of the input is usually simpler than having to specifically generate input with that property.

File: test/Test/Database/LSMTree/Internal/RunReader.hs [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+import qualified Database.LSMTree.Internal.WriteBuffer as WB
+import qualified System.FS.API as FS
+import qualified System.FS.BlockIO.API as FS
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck
+import           Test.Util.FS (noOpenHandles, withSimHasBlockIO,
+                     withTempIOHasBlockIO)
+import           Test.Util.Orphans ()
+
+-- TODO: this is pretty redundant with the tests for 'Run.Run'. Deduplicate?
```
Comment (mheinzel) [inline test/Test/Database/LSMTree/Internal/RunReader.hs:32]: Has been de-duplicated, so the TODO can be dropped.

File: test/Test/Database/LSMTree/Internal/Run.hs (hunk 11 above)
Comment (mheinzel) [reply test/Test/Database/LSMTree/Internal/Run.hs]: There are tests for this now: https://github.com/IntersectMBO/lsm-tree/pull/346/files#diff-48a02884e73d81f478092b7cb891f5af973b8b6844257c98f597415d0652e159

**Review by mheinzel — COMMENTED** [review]
(seems like I accidentally submitted parts of the review already)

The `RawPage` tests still seem a bit more complicated than they could be (see comments about generator, also [comparing against a reference implementation](https://github.com/IntersectMBO/lsm-tree/pull/346#discussion_r1742246587) could be simpler than the three separate properties?). This doesn't necessarily have to be a blocker, though.

However, it seems like `rawPageLookup` now allocates more, which should be investigated.

**Comment by recursion-ninja** [comment]
## Regression Fix Report

I have isolated and corrected the allocation regression in `rawPageLookup` that the PR contained. The excess allocation came from the "linear scanning" portion of the `bisectPageToKey` function. 

Previously, when the bench-marking showed identical allocations and comparable run-times, there was a defect in the `bisectPageToKey` function. To correct the defect, an additional check was required to determine of the binary search ended at the last entry in the page. If so, the binary search result needed to be changed from "found closest key at index `i`" to "key not found in page." This additional comparison work, which existed in the terminal "linear scan" portion of the binary search routine, was inexorably linked to the additional allocation. 

I attempted to alter the code while retaining the linear scan portion of the binary search routine. After repeated an persistent failure to remove the additional allocation, I removed the linear scan segment of the binary search routine.  To my pleasant surprise, the allocations decreased to the same *or less than* the allocations reported on `main`!

Below is a benchmark table I collated describing the change in bytes allocated (B) and run-time measured (ns) between `main` and this PR across recent versions of GHC. In the Δ column, a negative number represents an *improvement*  when measuring the performance of this PR (relative to `main`).

## Results

| `rawPageLookup`  | GHC        | `main` (B, ns)    | PR-#346 (B, ns)   | Diif-Δ (B, ns)    |
|:-----------------|:-----------|------------------:|------------------:|------------------:|
| `missing`        | **9.10.1** | (`544`, `101.50`) | (`544`, `104.90`) | (`± 0`, `+ 3.40`) |
| `existing-head`  | **9.10.1** | (`560`, ` 98.72`) | (`560`, `101.70`) | (`± 0`, `+ 2.98`) |
| `existing-last`  | **9.10.1** | (`640`, ` 96.91`) | (`576`, ` 88.27`) | (`-64`, `- 8.64`) |
| `missing`        | **9.8.2**  | (`544`, `102.90`) | (`544`, `104.90`) | (`± 0`, `+ 2.00`) |
| `existing-head`  | **9.8.2**  | (`560`, `100.40`) | (`560`, `103.00`) | (`± 0`, `+ 2.60`) |
| `existing-last`  | **9.8.2**  | (`640`, ` 98.92`) | (`576`, ` 91.17`) | (`-64`, `- 7.75`) |
| `missing`        | **9.6.6**  | (`544`, `104.20`) | (`544`, `103.50`) | (`± 0`, `- 0.70`) |
| `existing-head`  | **9.6.6**  | (`560`, ` 99.82`) | (`560`, ` 90.30`) | (`± 0`, `- 9.52`) |
| `existing-last`  | **9.6.6**  | (`640`, ` 97.21`) | (`576`, ` 88.48`) | (`-64`, `- 9.73`) |
| `missing`        | **9.4.8**  | (`544`, `102.90`) | (`544`, `105.40`) | (`± 0`, `+ 2.50`) |
| `existing-head`  | **9.4.8**  | (`560`, `100.00`) | (`560`, `103.10`) | (`± 0`, `+ 3.10`) |
| `existing-last`  | **9.4.8**  | (`640`, ` 99.06`) | (`576`, ` 88.84`) | (`-64`, `-10.22`) |

## Conclusion

My reading of the tabulated measurements is that the PR, with the recent regression fix, now constitutes an *unequivocal improvement* in **allocations** and a *minor to negligible improvement* in **run-time**!

**Comment by recursion-ninja** [comment]
I believe that this resolves the last critical technical issue of the PR. I'm going to give the PR a final review to ensure that all comments have been addressed.

File: test/Test/Database/LSMTree/Internal/RawPage.hs (hunk 21 above)
Comment (recursion-ninja) [reply test/Test/Database/LSMTree/Internal/RawPage.hs:464]: I took your advice to just simplify the generator and change the distribution to choose an existing key with $\mathbf{P}\left\[\frac{1}{2}\right\]$ and an `arbitrary` key otherwise.

Comment (recursion-ninja) [reply test/Test/Database/LSMTree/Internal/RawPage.hs:464]: Actually, I've reversed my position in light of new information. I annotated the tests with [`label`](https://hackage.haskell.org/package/tasty-quickcheck/docs/Test-Tasty-QuickCheck.html#v:label) to see how the 50/50 distribution of random/existing key selection panned out. There was *essentially* no coverage of the "before page" and "after page" cases. So I decided to bring back the computationally more complicated, but more predictable distribution of outcomes.

After bringing back the original, more complex distribution, I updated the sub-generators to have an expanded value range. The updated ranges address the previously "weird" location-independent byte generation which you had commented on. 

I had chosen that admittedly weird generation technique to simplify the generation process. To correctly generate any key within an upper and/or lower byte-stream, I would need to track if I had generated a smaller/larger byte earlier in the sequence. This required dragging state through the QuickCheck generator which I had lazily avoided doing. That state dragging has been added now.

Comment (mheinzel) [reply test/Test/Database/LSMTree/Internal/RawPage.hs]: The usage of `min`/`max` still doesn't seem 100% right, but it should be good enough and lead to a good distribution of test cases.

**Review by mheinzel — APPROVED** [review]
Good working on tracking down the allocation regression. I think everything has been addressed now.
